client/src/details/UrlBuilder.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "UrlBuilder.h"
+
+#include "ErrorHandling.h"
+#include "ReportingHandler.h"
+
+#include <curl/curl.h>
+
+#define THROW_IF_CURLU_ERROR(curlCall, error)                                                                          \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        auto __curlUrlCode = (curlCall);                                                                               \
+        THROW_CODE_IF_NOT_LOG(error, __curlUrlCode == CURLUE_OK, m_handler, GetCurlUrlStrError(__curlUrlCode));        \
+    } while ((void)0, 0)
+
+#define THROW_IF_CURLU_SETUP_ERROR(curlCall) THROW_IF_CURLU_ERROR(curlCall, ConnectionUrlSetupFailed)

```

Comment:
In that case we'd either lose the ability to map the errors to strings, or have to implement in regards to every curl error.
Or even be too attached to curl in future in case we want to change the inner library.

I don't think there is a problem with mapping the URL errors to a single error. The message is good enough to distinguish scheme, host, whatever part is wrong. And they are all related to url setup error.

Reply:
Fair points.